packages/react-vapor/src/components/validation/hoc/WithDirtyCheckboxHOC.tsx

```diff
@@ -0,0 +1,53 @@
+import * as React from 'react';
+import {connect} from 'react-redux';
+
+import {IDispatch} from '../../../utils/ReduxUtils';
+import {ICheckboxOwnProps} from '../../checkbox/Checkbox';
+import {IInputOwnProps} from '../../input/Input';
+import {ValidationActions} from '../ValidationActions';
+import {ValidationTypes} from '../ValidationTypes';
+import {InferableComponentEnhancer} from './connectHOC';
+
+export interface IWithDirtyCheckboxOwnProps {
+    resetDirtyOnUnmount?: boolean;
+}
+
+const mapDispatchToProps = (dispatch: IDispatch) => ({
+    setIsDirty: (id: string, isDirty: boolean) =>
+        dispatch(ValidationActions.setDirty(id, isDirty, ValidationTypes.wrongInitialValue)),
+    clearIsDirty: (id: string) => dispatch(ValidationActions.clearDirty(id, ValidationTypes.wrongInitialValue)),
+});
+
+export const withDirtyCheckboxHOC = <T extends ICheckboxOwnProps & IInputOwnProps>(
+    Component: React.ComponentType<T>
+) => {
+    type DispatchProps = ReturnType<typeof mapDispatchToProps>;
+    const WrappedCheckbox: React.FunctionComponent<T & IWithDirtyCheckboxOwnProps & DispatchProps> = ({
+        setIsDirty,
+        clearIsDirty,
+        handleOnClick,
+        resetDirtyOnUnmount,
+        ...props
+    }) => {
+        React.useEffect(
+            () => () => {
+                resetDirtyOnUnmount && clearIsDirty(props.id);
+            },
+            []
+        );
+
+        return (
+            <Component
+                {...(props as T)}
+                handleOnClick={(wasChecked: boolean) => {
+                    const isNowChecked = !wasChecked;
+                    setIsDirty(props.id, isNowChecked !== (props.defaultChecked || false));
```

Comment:
```suggestion
                    setIsDirty(props.id, isNowChecked !== props.defaultChecked);
```
```ts
(true || false) // simplifies to true
(false || false) // simplifies to false
```

Reply:
I felt the purpose was clearer, but OK :shrug:

Reply:
HEY the unit test is failing because of this, you forgot one use case, where `props.defaultChecked` is **undefined**. 

`true !== undefined` returns true
`false !== undefined` returns true

whereas

`true !== (undefined || false)` returns true.
`false !== (undefined || false)` **returns false**